rasa/core/policies/ted_policy.py

```diff
@@ -0,0 +1,642 @@
+import copy
+import logging
+import os
+import pickle
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_addons as tfa
+
+from typing import Any, List, Optional, Text, Dict, Tuple, Union
+
+import rasa.utils.io
+from rasa.core.domain import Domain
+from rasa.core.featurizers import (
+    TrackerFeaturizer,
+    FullDialogueTrackerFeaturizer,
+    LabelTokenizerSingleStateFeaturizer,
+    MaxHistoryTrackerFeaturizer,
+)
+from rasa.core.policies.policy import Policy
+from rasa.core.constants import DEFAULT_POLICY_PRIORITY, DIALOGUE
+from rasa.core.trackers import DialogueStateTracker
+from rasa.utils import train_utils
+from rasa.utils.tensorflow import layers
+from rasa.utils.tensorflow.transformer import TransformerEncoder
+from rasa.utils.tensorflow.models import RasaModel
+from rasa.utils.tensorflow.model_data import RasaModelData, FeatureSignature
+from rasa.utils.tensorflow.constants import (
+    LABEL,
+    HIDDEN_LAYERS_SIZES,
+    TRANSFORMER_SIZE,
+    NUM_TRANSFORMER_LAYERS,
+    NUM_HEADS,
+    MAX_SEQUENCE_LENGTH,
+    BATCH_SIZES,
+    BATCH_STRATEGY,
+    EPOCHS,
+    RANDOM_SEED,
+    RANKING_LENGTH,
+    LOSS_TYPE,
+    SIMILARITY_TYPE,
+    NUM_NEG,
+    EVAL_NUM_EXAMPLES,
+    EVAL_NUM_EPOCHS,
+    NEGATIVE_MARGIN_SCALE,
+    REGULARIZATION_CONSTANT,
+    SCALE_LOSS,
+    USE_MAX_NEG_SIM,
+    MAX_NEG_SIM,
+    MAX_POS_SIM,
+    EMBEDDING_DIMENSION,
+    DROP_RATE_DIALOGUE,
+    DROP_RATE_LABEL,
+    DROP_RATE_ATTENTION,
+    KEY_RELATIVE_ATTENTION,
+    VALUE_RELATIVE_ATTENTION,
+    MAX_RELATIVE_POSITION,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class TEDPolicy(Policy):
+    """Transformer Embedding Dialogue (TED) Policy.
+
+    The policy used in our paper https://arxiv.org/abs/1910.00486.
+    """
+
+    SUPPORTS_ONLINE_TRAINING = True
+
+    # default properties (DOC MARKER - don't remove)
+    defaults = {
+        # nn architecture
+        # a list of hidden layers sizes before dialogue and action embedding layers
+        # number of hidden layers is equal to the length of this list
+        HIDDEN_LAYERS_SIZES: {DIALOGUE: [], LABEL: []},
```

Comment:
I'd either use these constants to a module where it's clear that it relates to layers or I'd call therm `DIALOGUE_EMBEDDING_LAYER` and `LABEL_EMBEDDING_LAYER`. Because `DIALOGUE` in `rasa.constants` is quite confusing and also in here it's only understandable because of the comment

Reply:
they are not layers

Reply:
Should we name them `x_FEATURES` or `x_NAME` or `x_FEATURE_NAME`?

Reply:
it's not really features, it's the parameters of nn responsible for dialogue and label sides

Reply:
> they are not layers

But the comment above it says `layers`? 

> it's not really features, it's the parameters of nn responsible for dialogue and label sides

How about `x_PARAMETER_SIZES`?

Reply:
it's already says sizes in `HIDDEN_LAYERS_SIZES`, why would you make things longer?